src/apps/lwaftr/binding_table.lua

```diff
@@ -1,106 +1,469 @@
+      end
+      parser:skip_whitespace()
+      if parser:check(',') or parser:check(';') then
+         parser:skip_whitespace()
+      end
+   end
+   return builder:build()
+end
+
+local function parse_br_addresses(parser)
+   local addresses = {}
+   parser:skip_whitespace()
+   parser:consume_token('[%a_]', 'br_addresses')
+   parser:skip_whitespace()
+   parser:consume('{')
+   parser:skip_whitespace()
+   while not parser:check('}') do
+      table.insert(addresses, parser:parse_ipv6())
+      parser:skip_whitespace()
+      if parser:check(',') then parser:skip_whitespace() end
+   end
+   if #addresses == 0 then parser:error('no lwaftr addresses specified') end
+   local ret = ffi.new(ffi.typeof('$[?]', br_address_t), #addresses)
+   for i, addr in ipairs(addresses) do ret[i-1].addr = addr end
+   return ret, #addresses
+end
+
+local function parse_softwires(parser, psid_map, br_address_count)
+   local function required(key)
+      return function(config)
+         error('missing required configuration key "'..key..'"')
       end
-      if not inserted[pv4] then
-         inserted[pv4] = {psid_len = psid_len}
+   end
+   local softwire_spec = {
+      parse={
+         ipv4=Parser.parse_ipv4_as_uint32,
+         psid=Parser.parse_psid,
+         b4=Parser.parse_ipv6,
+         aftr=Parser.parse_non_negative_number
+      },
+      defaults={
+         ipv4=required('ipv4'),
+         psid=function(config) return 0 end,
+         b4=required('b4'),
+         aftr=function(config) return 0 end
+      },
+      validate=function(parser, config)
+         local psid_length = psid_map:lookup(config.ipv4).value.psid_length
+         if config.psid >= 2^psid_length then
+            parser:error('psid %d out of range for IP', config.psid)
+         end
+         if config.aftr >= br_address_count then
+            parser:error('only %d br addresses are defined', br_address_count)
+         end
       end
-      b4_v6 = ipv6:pton(b4_v6)
-      local pentry
-      if br_v6 then
-         pentry = {b4_v6, pv4, psid_info, ipv6:pton(br_v6)}
-      else
-         pentry = {b4_v6, pv4, psid_info}
+   }
+
+   local map = phm.PodHashMap.new(softwire_key_t, softwire_value_t,
+                                  hash_softwire)
+   local key, value = softwire_key_t(), softwire_value_t()
+   parser:skip_whitespace()
+   parser:consume_token('[%a_]', 'softwires')
+   parser:skip_whitespace()
+   parser:consume('{')
+   parser:skip_whitespace()
+   while not parser:check('}') do
+      local entry = parser:parse_property_list(softwire_spec, '{', '}')
+      key.ipv4, key.psid = entry.ipv4, entry.psid
+      value.br, value.b4_ipv6 = entry.aftr, entry.b4
+      map:add(key, value)
+      local success = true
+      --local success = pcall(map.add, map, key, value)
+      if not success then
+         parser:error('duplicate softwire for ipv4=%s, psid=%d',
+                      lwdebug.format_ipv4(key.ipv4), key.psid)
```

Comment:
It seems like originally `map.add` can notify whether an entry was added successfully or not, but the line is commented and `success` is always true, so the error is never going to be reported. If that's what is intended, I would remove the commented line and add a TODO comment. How does it sound?